### PR TITLE
New functions to read elf files and parse modules (SP3)

### PR DESCRIPTION
**Pull request by softwarecki**
Rewritten elf file reading and module parsing:
-  **Set of new functions for reading elf files.** A structure was created to represent a file, a section and a strings section. Created a new functions to read a elf file, retrieve a section header based on index or name, read a section contents based on a header or name, retrieve section name and functions that print file, section and program headers.
- **Set of a new functions to parse modules.** A structure was created to represent the module, module section and section
informations. Added a set of new functions for parsing a module and reading its contents.

Moved the module information to the new module structure. Removed elf.c file that is no longer in use.

SOF CI PR https://github.com/thesofproject/sof/pull/7413

### CODE REVIEW

**Comment by marc-hb** [comment]
I don't understand, there are `elfutils` libraries already, why are we re-inventing this? Parsing binaries in C in 2023 looks like instant technical debt.

File: src/module.c [hunk 1]
```diff
@@ -0,0 +1,503 @@
...
+{
+	uint32_t a_start = a->address;
+	uint32_t a_end = a_start + a->size;
+
+	uint32_t b_start = b->address;
+	uint32_t b_end = b_start + b->size;
+
+	/* is section start overlapping ? */
+	return (a_start >= b_start && a_start < b_end) ||
+	/* is section end overlapping ? */
```
Comment (lgirdwood) [inline src/module.c:445]: indentation

**Review by lgirdwood — COMMENTED** [review]
LGTM.

**Comment by lgirdwood** [comment]
> I don't understand, there are `elfutils` libraries already, why are we re-inventing this? Parsing binaries in C in 2023 looks like instant technical debt.

Probably not as simple to use elfutils for the required usage and nothing wrong with C here.

**Comment by plbossart** [comment]
> > I don't understand, there are `elfutils` libraries already, why are we re-inventing this? Parsing binaries in C in 2023 looks like instant technical debt.
> 
> Probably not as simple to use elfutils for the required usage and nothing wrong with C here.

It's however a legitimate question to ask where this code comes from and how it was validated, no? Please tell me it didn't come from ChatGPT :-)

**Comment by lgirdwood** [comment]
@softwarecki any update ?

**Comment by softwarecki** [comment]
@lgirdwood I'll try to finish work on this this PR soon.

**Comment by marc-hb** [comment]
Why parsing binaries in C in 2023 is a waste of time:
- https://github.com/thesofproject/sof/pull/8216

**Comment by softwarecki** [comment]
@plbossart: The code is the result of human work and was not created by chatGPT. The latter has the ability to give elaborate answers without any substance and sometimes has hallucinations. To perform a validation, I created a PR in SOF https://github.com/thesofproject/sof/pull/7413.

@lgirdwood I think its ready now :)

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
@kv2019i fyi.

**Comment by marc-hb** [comment]
> There are elfutils libraries already, why are we re-inventing this? Parsing binaries in C in 2023 looks like instant technical debt.

Ping, @softwarecki please explain how is this not re-inventing https://www.freshports.org/devel/elfutils/

**Comment by softwarecki** [comment]
@marc-hb I concluded that it would be better to write a few simple functions for reading files than add a dependency to another (quite large) library. If this is a problem for you, you can use `elfutils` and create your own PR. It will be easier now, as I separated all elf file operations into a separate module.

**Review by lgirdwood — APPROVED** [review]

**Review by kv2019i — APPROVED** [review]
Given this is replacing in-tree elf parser with another one and author of old code is approving, I'm good to go with this.

**Pull request merged**